policy/src/main/java/org/apache/brooklyn/policy/action/StopAfterDurationPolicy.java

```diff
@@ -0,0 +1,135 @@
+            .defaultValue(Lifecycle.RUNNING)
+            .constraint(Predicates.notNull())
+            .build();
+
+    public static final ConfigKey<Duration> POLL_PERIOD = ConfigKeys.builder(Duration.class)
+            .name("pollPeriod")
+            .description("Period in which duration-since sensor should be updated")
+            .defaultValue(Duration.THIRTY_SECONDS)
+            .constraint(Predicates.notNull())
+            .build();
+
+    public static final ConfigKey<Boolean> HAS_STARTED_TIMER = ConfigKeys.builder(Boolean.class)
+            .name("timer-started")
+            .description("For internal use only")
+            .defaultValue(false)
+            .build();
+
+    public static final ConfigKey<Boolean> FIRED_STOP = ConfigKeys.builder(Boolean.class)
+            .name("fired-stop")
+            .description("For internal use only")
+            .defaultValue(false)
+            .build();
+
+    private final Object eventLock = new Object[0];
+
+    public void setEntity(final EntityLocal entity) {
+        super.setEntity(entity);
+        entity.subscriptions().subscribe(entity, Attributes.SERVICE_STATE_ACTUAL, new LifecycleListener());
+        entity.subscriptions().subscribe(entity, Sensors.newSensor(Duration.class, getSensorName()), new TimeIsUpListener());
+    }
+
+    @Override
+    protected <T> void doReconfigureConfig(ConfigKey<T> key, T val) {
+        Set<ConfigKey<?>> accepted = ImmutableSet.<ConfigKey<?>>of(
+                HAS_STARTED_TIMER,
+                FIRED_STOP,
+                LIFETIME);
+        if (!accepted.contains(key)) {
+            super.doReconfigureConfig(key, val);
+        }
+    }
+
+    private String getSensorName() {
+        return "duration.since.first-" + config().get(STATE).name().toLowerCase();
+    }
+
+    private class LifecycleListener implements SensorEventListener<Lifecycle> {
+        @Override
+        public void onEvent(SensorEvent<Lifecycle> event) {
+            synchronized (eventLock) {
+                if (!config().get(HAS_STARTED_TIMER) && config().get(STATE).equals(event.getValue())) {
+                    DurationSinceSensor sensor = new DurationSinceSensor(ConfigBag.newInstance(ImmutableMap.of(
+                            DurationSinceSensor.SENSOR_NAME, getSensorName(),
+                            DurationSinceSensor.SENSOR_PERIOD, config().get(POLL_PERIOD),
+                            DurationSinceSensor.SENSOR_TYPE, Duration.class.getName())));
+                    sensor.apply(entity);
+                    config().set(HAS_STARTED_TIMER, true);
+                }
+            }
+        }
+    }
+
+    private class TimeIsUpListener implements SensorEventListener<Duration> {
+        @Override
+        public void onEvent(SensorEvent<Duration> event) {
+            synchronized (eventLock) {
+                if (!config().get(FIRED_STOP)) {
+                    if (!config().get(LIFETIME).subtract(event.getValue()).isPositive()) {
```

Comment:
It's a niggle, but is there an "isNegative" method and would this be clearer:
`config().get(LIFETIME).subtract(event.getValue()).isNegative()`